metagraph/src/graph/representation/succinct/dbg_succinct.cpp

```diff
@@ -164,34 +164,43 @@ void DBGSuccinct::call_incoming_kmers(node_index node,
 
 void DBGSuccinct::adjacent_outgoing_nodes(node_index node,
                                           const std::function<void(node_index)> &callback) const {
-    assert(node > 0 && node <= num_nodes());
+    assert(is_valid(node));
 
-    call_outgoing(*boss_graph_, kmer_to_boss_index(node), [&](auto i) {
-        auto next = boss_to_kmer_index(i);
-        if (next != npos)
+    call_outgoing(*boss_graph_, node, [&](auto i) {
+        auto next = i;
+        if (is_valid(next))
             callback(next);
     });
 }
 
 void DBGSuccinct::adjacent_incoming_nodes(node_index node,
                                           const std::function<void(node_index)> &callback) const {
-    assert(node > 0 && node <= num_nodes());
+    assert(is_valid(node));
 
-    auto edge = kmer_to_boss_index(node);
+    auto edge = node;
 
     boss_graph_->call_incoming_to_target(boss_graph_->bwd(edge),
         boss_graph_->get_node_last_value(edge),
         [&](BOSS::edge_index incoming_boss_edge) {
             assert(boss_graph_->get_W(incoming_boss_edge) % boss_graph_->alph_size
                     == boss_graph_->get_node_last_value(edge));
 
-            auto prev = boss_to_kmer_index(incoming_boss_edge);
-            if (prev != npos)
+            auto prev = incoming_boss_edge;
+            if (is_valid(prev))
                 callback(prev);
         }
     );
 }
 
+void DBGSuccinct::call_nodes(const std::function<void(node_index)> &callback,
+                             const std::function<bool()> &terminate) const {
+    for (node_index i = 1; i <= max_index() && !terminate(); ++i) {
+        if (is_valid(i)) {
+            callback(i);
+        }
+    }
```

Comment:
I generally dislike the idea of using exceptions for control flow. I implemented this approach for now, but I feel like there should be other ways to work with it. For example, encode termination condition with the return value (instead of `void`). But it seems like it will require another huge overhaul of the codebase...

On the same note, it seems that some functions use `terminate()` for this, and others `stop_early()` :pensive: